### PR TITLE
Feature/collection field

### PR DESCRIPTION
**Pull request by NathHorrigan**
## The Basics

The collection field allows more control in add Lists to your graphql types, It follows a more composable syntax that makes more sense to me then just passing a `is_list=True` param each type. Instead you have a `GraphQLCollection` field which wraps another type (such as `GraphQLString`) in the same way you'd wrap an array in `[]`. 

`GraphQLForeignKey("related_links", "home.blogpagerelatedlink", is_list=True),` becomes:

```
GraphQLCollection(
    GraphQLForeignKey,
    "related_links",
    "home.blogpagerelatedlink"
),
```

## ForeignKey field extraction:

Another feature implemented in the Collection API is the ability to iterate over a list of foreign keys and return only one field from each model like so (building on previous example, look at the source param):

```
GraphQLCollection(
    GraphQLString,
    "related_urls",
    source="related_links.url"
),
```

A query like so now returns:

Query:
```
{
  pages {
    ...on BlogPage {
      relatedLinks {
        id
        name
        url
      }
      relatedUrls
    }
  }
}
```

Result:
```

        "relatedLinks": [
          {
            "id": "1",
            "name": "Home",
            "url": "http://localhost:8000/home/"
          },
          {
            "id": "2",
            "name": "About",
            "url": "http://localhost:8000/about/"
          },
          {
            "id": "3",
            "name": "Team",
            "url": "http://localhost:8000/team/"
          },
          {
            "id": "4",
            "name": "Blog",
            "url": "http://localhost:8000/blog/"
          }
        ],
        "relatedUrls": [
          "Home",
          "About",
          "Team",
          "Blog"
        ]
      }
```

### CODE REVIEW

File: grapple/utils.py [hunk 1]
```diff
@@ -49,6 +49,7 @@ def resolve_queryset(
     if order is not None:
         qs = qs.order_by(*map(lambda x: x.strip(), order.split(",")))
 
+    qs = qs.all()
```
Comment (zerolab) [inline grapple/utils.py]: why is this needed here?

File: grapple/actions.py [hunk 2]
```diff
@@ -115,9 +115,47 @@ def get_fields_and_properties(cls):
...
+
+def model_resolver(field):
     def mixin(self, instance, info, **kwargs):
-        return getattr(instance, field_source)
+        from .utils import resolve_queryset
+
+        cls_field = getattr(instance, field.field_source)
+
+        # If queryset then call .all() method
+        if hasattr(cls_field, "all"):
```
Comment (zerolab) [inline grapple/actions.py]: are there cases where there'd be an `all` attribute, but this is not a queryset?

Comment (NathHorrigan) [reply grapple/actions.py]: Now use: `if issubclass(type(cls_field), models.Manager):`

File: grapple/models.py [hunk 3]
```diff
@@ -66,71 +68,98 @@ def Mixin():
...
     def Mixin():
-        from .types.media import MediaObjectType
-
-        return GraphQLField(field_name, MediaObjectType)
+        from .types.structures import QuerySetList
+
+        # Check if using nested field extracion:
+        source = kwargs.get("source", None)
+        if source and "." in source:
+            source, key = source.split(".")
```
[the remark this answers is not in the document]
Comment (NathHorrigan) [reply grapple/models.py]: Multi level support addeds can now do:
```
       GraphQLCollection(
            GraphQLString,
            "authors",
            source="authors.person.name"
        ),
```

File: example/home/models.py [hunk 4]
```diff
@@ -89,7 +90,16 @@ def copy(self):
...
-        GraphQLForeignKey("related_links", "home.blogpagerelatedlink", True),
+        GraphQLCollection(
+            GraphQLForeignKey,
+            "related_links",
+            "home.blogpagerelatedlink"
+        ),
+        GraphQLCollection(
+            GraphQLString,
+            "related_urls",
+            source="related_links.name"
```
[the remark this answers is not in the document]
Comment (NathHorrigan) [reply example/home/models.py]: Done!

**Comment by zerolab** [comment]
It reads much better now. Just needs tests 👍

File: docs/general-usage/model-types.rst [hunk 5]
```diff
@@ -45,6 +45,67 @@ GraphQLString
         }
 
 
+GraphQLCollection
+-------------
+.. module:: grapple.models
+.. class:: GraphQLCollection(nested_type, *args, **kwargs)
+
+    A field type that wraps another model type in a list, Most likely to be used when referencing ForeignKey lists (like when using Orderables).
```
Comment (zerolab) [inline docs/general-usage/model-types.rst]: "A field type that wraps another model type in a list. Best suited for referencing `Orderable`s (i.e. `ForeignKey` lists)."

File: docs/general-usage/model-types.rst [hunk 6]
```diff
@@ -45,6 +45,67 @@ GraphQLString
...
+
+    .. attribute:: nested_type
+
+        Pass another Grapple model type such as `GraphQLString` or `GraphQLForeignKey`.
+
+    .. attribute:: *args
+
+        Any args that you want to pass on to the nested type, you will always be passing `field_name` here for example.
```
[the remark this answers is not in the document]
Comment (zerolab) [reply docs/general-usage/model-types.rst]: also, if `field_name` must be always passed, should it not be added to method signature?

File: docs/general-usage/model-types.rst [hunk 7]
```diff
@@ -45,6 +45,67 @@ GraphQLString
...
+
+    .. attribute:: **kwargs
+
+        Any keyword args that you want to pass on to the nested type. 
+        
+        One keyword argument that is more powerful with Collections is the `source` argument. With ``GraphQLCollection``, 
+        you can pass a source string that is multiple layers deep and Grapple will handle the querying for you through
```
Comment (zerolab) [inline docs/general-usage/model-types.rst]: ", you can pass" -> ". You can pass"

Comment (NathHorrigan) [reply docs/general-usage/model-types.rst]: Done.

File: docs/general-usage/model-types.rst (hunk 5 above)
Comment (NathHorrigan) [reply docs/general-usage/model-types.rst]: Done.

File: docs/general-usage/model-types.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (NathHorrigan) [reply docs/general-usage/model-types.rst]: Signature Changed.
